pkg/i2gw/providers/istio/resource_reader.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package istio
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type reader struct {
+	k8sClient client.Client
+}
+
+func newResourceReader(k8sClient client.Client) reader {
+	return reader{
+		k8sClient: k8sClient,
+	}
+}
+
+func (r *reader) fetchResourcesFromCluster(ctx context.Context) (*storage, error) {
+	res := newResourcesStorage()
+
+	gateways, err := r.readGatewaysFromCluster(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read gateways: %w", err)
+	}
+
+	res.Gateways = gateways
+
+	virtualServices, err := r.readVirtualServicesFromCluster(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read virtual services: %w", err)
+	}
+
+	res.VirtualServices = virtualServices
+
+	return &res, nil
+}
+
+func (r *reader) readUnstructuredObjects(objects []*unstructured.Unstructured) (*storage, error) {
```

Comment:
What's the purpose of this function? As far as I see, it is unused

Reply:
This func will be used to fetch all required Istio CRDs. Now not used as this PR just prepared the code to make actual Istio conversion. But this will be updated in future PRs.